pkg/router/skipper.go

```diff
@@ -0,0 +1,235 @@
+package router
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+
+	flaggerv1 "github.com/weaveworks/flagger/pkg/apis/flagger/v1beta1"
+)
+
+/*
+Skipper Principles:
+* if only one backend has a weight, only one backend will get 100% traffic
+* if two of three or more backends have a weight, only those two should get traffic.
+* if two backends don't have any weight, it's undefined and right now they get equal amount of traffic.
+* weights can be int or float, but always treated as a ratio.
+
+Implementation:
+* apex Ingress is immutable
+* new canary Ingress contains two paths for primary and canary service
+* canary Ingress manages weights on primary & canary service, hence no traffic to apex service
+
+*/
+
+const (
+	skipperpredicateAnnotationKey      = "zalando.org/skipper-predicate"
+	skipperBackendWeightsAnnotationKey = "zalando.org/backend-weights"
+	canaryPatternf                     = "%s-canary"
+	canaryRouteWeight                  = "Weight(100)"
+	canaryRouteDisable                 = "False()"
+)
+
+type SkipperRouter struct {
+	kubeClient kubernetes.Interface
+	logger     *zap.SugaredLogger
+}
+
+// Reconcile creates or updates the ingresses
+func (skp *SkipperRouter) Reconcile(canary *flaggerv1.Canary) error {
+	if canary.Spec.IngressRef == nil || canary.Spec.IngressRef.Name == "" {
+		return fmt.Errorf("ingress selector is empty")
+	}
+
+	apexSvcName, primarySvcName, canarySvcName := canary.GetServiceNames()
+	apexIngressName, canaryIngressName := skp.getIngressNames(canary.Spec.IngressRef.Name)
+
+	// retrieving apex ingress
+	apexIngress, err := skp.kubeClient.NetworkingV1beta1().Ingresses(canary.Namespace).Get(
+		context.TODO(), apexIngressName, metav1.GetOptions{})
```

Comment:
Should context.TODO() be replaced?

Reply:
it should, but that context is used everywhere in the project. Hard to make a call from my side for this project, thoughts @stefanprodan which context to choose?

Reply:
The TODO context was added when upgrading client-go to 1.16. I'll leave it like this for now and change it in different PR>

Reply:
thanks!